prefect_github/repository.py

```diff
@@ -23,6 +31,121 @@
+                        "credential if you wish to use the 'SSH' format and are not "
+                        "using a private repository, or you must change the repository "
+                        "url to the 'HTTPS' format. "
+                    )
+                )
+
+        return v
+
+    def _create_repo_url(self) -> str:
+        """Format the URL provided to the `git clone` command.
+
+        For private repos: https://<oauth-key>@github.com/<username>/<repo>.git
+        All other repos should be the same as `self.repository`.
+        """
+
+        if self.repository.startswith("https://") and self.credentials is not None:
+            repo_url = self.repository[8:]
+            token_value = self.credentials.token.get_secret_value()
+            full_url = f"https://{token_value}@{repo_url}"
+        else:
+            full_url = self.repository
+
+        return full_url
+
+    @staticmethod
+    def _get_paths(
+        dst_dir: Union[str, None], src_dir: str, sub_directory: str
+    ) -> Tuple[str, str]:
+        """Returns the fully formed paths for GitHubRepository contents in the form
+        (content_source, content_destination).
+        """
+        if dst_dir is None:
+            content_destination = Path(".").absolute()
+        else:
+            content_destination = Path(dst_dir)
+
+        content_source = Path(src_dir)
+
+        if sub_directory:
+            content_destination = content_destination.joinpath(sub_directory)
+            content_source = content_source.joinpath(sub_directory)
+
+        return str(content_source), str(content_destination)
+
+    @sync_compatible
+    async def get_directory(
+        self, from_path: Optional[str] = None, local_path: Optional[str] = None
+    ) -> None:
+        """
+        Clones a GitHub project specified in `from_path` to the provided `local_path`;
+        defaults to cloning the repository reference configured on the Block to the
+        present working directory.
+
+        Args:
+            from_path: If provided, interpreted as a subdirectory of the underlying
+                repository that will be copied to the provided local path.
+            local_path: A local path to clone to; defaults to present working directory.
+        """
+        # CONSTRUCT COMMAND
+        cmd = f"git clone {self._create_repo_url()}"
+        if self.reference:
+            cmd += f" -b {self.reference} --depth 1"
+
+        # Clone to a temporary directory and move the subdirectory over
+        with TemporaryDirectory(suffix="prefect") as tmp_dir:
+            tmp_path_str = tmp_dir
+            cmd += f" {tmp_path_str}"
+
+            err_stream = io.StringIO()
+            out_stream = io.StringIO()
+            process = await run_process(cmd, stream_output=(out_stream, err_stream))
+            if process.returncode != 0:
+                err_stream.seek(0)
+                raise OSError(f"Failed to pull from remote:\n {err_stream.read()}")
```

Comment:
I think a `RuntimeError` makes more sense here, but it's not a strong conviction.